src/System.Text.Json/src/System/Text/Json/JsonElement.cs

```diff
@@ -0,0 +1,458 @@
+        }
+
+        public decimal GetDecimal()
+        {
+            if (TryGetDecimal(out decimal value))
+            {
+                return value;
+            }
+
+            throw new FormatException();
+        }
+
+        internal string GetPropertyName()
+        {
+            CheckValidInstance();
+
+            return _parent.GetNameOfPropertyValue(_idx);
+        }
+
+        public string GetRawText()
+        {
+            CheckValidInstance();
+
+            return _parent.GetRawValueAsString(_idx);
+        }
+
+        internal string GetPropertyRawText()
+        {
+            CheckValidInstance();
+
+            return _parent.GetPropertyRawValueAsString(_idx);
+        }
+
+        public ArrayEnumerator EnumerateArray()
+        {
+            CheckValidInstance();
+
+            JsonTokenType tokenType = TokenType;
+
+            if (tokenType != JsonTokenType.StartArray)
+            {
+                throw ThrowHelper.GetJsonElementWrongTypeException(JsonTokenType.StartArray, tokenType);
+            }
+
+            return new ArrayEnumerator(this);
+        }
+
+        public ObjectEnumerator EnumerateObject()
+        {
+            CheckValidInstance();
+
+            JsonTokenType tokenType = TokenType;
+
+            if (tokenType != JsonTokenType.StartObject)
+            {
+                throw ThrowHelper.GetJsonElementWrongTypeException(JsonTokenType.StartObject, tokenType);
+            }
+
+            return new ObjectEnumerator(this);
+        }
+
+        public override string ToString()
+        {
+            switch (TokenType)
+            {
+                case JsonTokenType.None:
+                case JsonTokenType.Null:
+                    return string.Empty;
+                case JsonTokenType.True:
+                    return bool.TrueString;
+                case JsonTokenType.False:
+                    return bool.FalseString;
+                case JsonTokenType.Number:
+                case JsonTokenType.StartArray:
+                case JsonTokenType.StartObject:
+                {
+                    // null parent should have hit the None case
+                    Debug.Assert(_parent != null);
+                    return _parent.GetRawValueAsString(_idx);
+                }
+                case JsonTokenType.String:
+                    return GetString();
+                case JsonTokenType.Comment:
+                case JsonTokenType.EndArray:
+                case JsonTokenType.EndObject:
+                default:
+                    Debug.Fail($"No handler for {nameof(JsonTokenType)}.{TokenType}");
+                    return string.Empty;
+            }
+        }
+
+        private void CheckValidInstance()
+        {
+            if (_parent == null)
+            {
+                throw new InvalidOperationException();
+            }
+        }
+
+        public struct ArrayEnumerator : IEnumerable<JsonElement>, IEnumerator<JsonElement>
```

Comment:
They're `public struct` which means they avoid an object allocation and virtual dispatch on MoveNext() and get_Current().  Most users don't see them because they'll write it into a foreach, which prefers pattern-foreach over IEnumerable-foreach (language rules).

Reply:
OK. Add [nonbrowsable]?

Reply:
We don't seem to have done that for other nested enumerators, for example https://source.dot.net/#System.Collections/System/Collections/Generic/LinkedList.cs,674b2c2e532e0349